packages/models-library/tests/test_user_preferences.py

```diff
@@ -0,0 +1,162 @@
+    return request.param
+
+
+def _get_base_user_preferences_data(
+    preference_type: PreferenceType, value: Any
+) -> dict[str, Any]:
+    return {"preference_type": preference_type, "value": value}
+
+
+def _get_utc_timestamp() -> float:
+    return arrow.utcnow().datetime.timestamp()
+
+
+@pytest.mark.parametrize("preference_type", PreferenceType)
+def test_base_user_preference_model(value: Any, preference_type: PreferenceType):
+    base_data = _get_base_user_preferences_data(
+        preference_type=preference_type, value=value
+    )
+    assert parse_obj_as(_BaseUserPreferenceModel, base_data)
+
+
+def test_backend_preferences(value: Any):
+    base_data = _get_base_user_preferences_data(
+        preference_type=PreferenceType.BACKEND, value=value
+    )
+    backend_preference = parse_obj_as(BaseBackendUserPreference, base_data)
+    # check serialization
+    assert set(backend_preference.dict().keys()) == {"value"}
+
+
+def test_frontend_preferences(value: Any):
+    base_data = _get_base_user_preferences_data(
+        preference_type=PreferenceType.FRONTEND, value=value
+    )
+
+    data_with_rendered_widget = deepcopy(base_data)
+    data_with_rendered_widget.update(
+        {
+            "preference_identifier": "pref-name",
+        }
+    )
+    # check serialization
+    frontend_preference = parse_obj_as(
+        BaseFrontendUserPreference, data_with_rendered_widget
+    )
+    assert set(frontend_preference.dict().keys()) == {"value"}
+
+
+@pytest.mark.parametrize("service_key", _SERVICE_KEY_SAMPLES)
+def test_user_service_preferences(value: Any, service_key: ServiceKey):
+    base_data = _get_base_user_preferences_data(
+        preference_type=PreferenceType.USER_SERVICE, value=value
+    )
+    base_data.update(
+        {
+            "service_key": service_key,
+            "last_changed_utc_timestamp": _get_utc_timestamp(),
+        }
+    )
+    instance = parse_obj_as(BaseUserServiceUserPreference, base_data)
+    assert set(instance.dict().keys()) == {"service_key", "value"}
+
+
+@pytest.fixture
+def unregister_defined_classes() -> Iterator[None]:
+    yield
+    # pylint: disable=protected-access
+    _AutoRegisterMeta._registered_user_preference_classes.pop(  # noqa: SLF001
+        "Pref1", None
+    )
+
+
+def test_user_defined_backend_preference(value: Any, unregister_defined_classes: None):
+    # definition of a new custom property
+    class Pref1(BaseBackendUserPreference):
+        ...
+
+    # usage
+    pref1 = Pref1(value=value)
+    assert isinstance(pref1, BaseBackendUserPreference)
+
+    # check bytes serialization/deserialization
+    pref1_as_bytes = pref1.json().encode()
+    new_instance = Pref1.parse_raw(pref1_as_bytes)
+    assert new_instance == pref1
+
+
+def test_user_defined_frontend_preference(value: Any, unregister_defined_classes: None):
+    # definition of a new custom property
+    class Pref1(BaseFrontendUserPreference):
+        preference_identifier = "pref1"
```

Comment:
ok so now I see how you use it, that is nice!
But that means, that each time the frontend adds a preference then it needs to change backend code as well. where directly storing as Json would not need that.

And second question, how will changes of preferences be covered? i.e. migrations. say a preference changes its type?